torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
   return getImmediateByType(poly->dtype(), GCD);
 }
 
-// Searches the polynomial for Terms that can be merged in the Round + Mod
-// pattern: (x/y) * y + x % y => RoundOff(x,y) + Mod(x, y) => x.
+// A ModRound is a div-mod-mul in which the divisor in div and multiplier in mul
+// are identical. Return (scalar, denominator, divisor, mod_divisor). Example:
+// ((t/7)%9)*7 -> (1, t, 7, 9)
+std::tuple<const Expr*, const Expr*, const Expr*, const Expr*> modRound(
+    const Term* e) {
+  const Div* div{nullptr};
+  const Mod* mod{nullptr};
+  const Expr* denom{nullptr};
+  const Expr* divisor = new IntImm(1);
+  const Expr* mod_divisor{nullptr};
+  const Expr* multiplier = e->scalar();
+  const Expr* scalar = new IntImm(1);
+  const Expr* other(nullptr);
+
+  for (auto* m : e->variables()) {
+    if (m->expr_type() == IRNodeType::kMod) {
+      // TODO: currently only identify terms with one variable being mod; it is
+      // possible to extend this if we have to handle terms like (t/(x%2 * y) %
+      // z) * (x%2 *y).
+      if (!mod) {
+        mod = dynamic_cast<const Mod*>(m);
+      } else {
+        return std::make_tuple(nullptr, nullptr, nullptr, nullptr);
+      }
+    } else {
+      // All non-mod vairables are considered as part of the multiplier.
+      multiplier = new Mul(multiplier, m);
+    }
+  }
+  multiplier = IRSimplifier::simplify(multiplier);
+
+  if (!mod) {
+    return std::make_tuple(nullptr, nullptr, nullptr, nullptr);
+  }
+
+  mod_divisor = IRSimplifier::simplify(mod->rhs());
+  other = mod->lhs();
+
+  if (!(div = dynamic_cast<const Div*>(other))) {
+    return std::make_tuple(nullptr, nullptr, nullptr, nullptr);
+  }
+
+  divisor = IRSimplifier::simplify(div->rhs());
+  other = div->lhs();
+
+  // Deny cases in which divisor=1.
+  if (divisor->isConstant() && immediateEquals(divisor, 1)) {
+    return std::make_tuple(nullptr, nullptr, nullptr, nullptr);
+  }
+
+  // Deny cases in which divisor!=multiplier.
+  HashProvider& hasher = e->hasher();
+  if (hasher.hash(divisor) != hasher.hash(multiplier)) {
+    if (divisor->isConstant() && multiplier->isConstant()) {
```

Comment:
Yeah, I think it would be great if we can extend it. 
As we discussed, there are currently no simplifications like `x*2/x -> 2` (because x could be 0) so I'm not able to extract `2` out of the `multiplier` to be `scalar`. But we can extend division to 2 versions: 1) division for customers that has to be strictly simplified  and 2) division we introduced in our transformations  which can be simplified without considering 0s, e.g., `Div_nonzero`. The second division will be only used to facilitate our transformations.

Reply:
Yeah, that sounds good! Please put a TODO in the code with these thoughts and we can return to that later.